examples/model_selection/plot_grid_search_refit_callable.py

```diff
@@ -0,0 +1,120 @@
+"""
+==================================================
+Balance model complexity and cross-validated score
+==================================================
+
+This example balances model complexity and cross-validated score by
+finding a decent accuracy within 1 standard deviation of the best accuracy
+score while minimising the number of PCA components. This is a rule of thumb
+for insignificant difference.
+
+The figure shows the trade-off between cross-validated score and number of
+PCA components. The balanced case is when n_components=6 and accuracy=0.77,
+which falls into the range within 1 standard deviation of the best accuracy
+score.
+"""
+# Author: Wenhao Zhang <wenhaoz@ucla.edu>
+
+print(__doc__)
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn.datasets import load_digits
+from sklearn.decomposition import PCA
+from sklearn.model_selection import GridSearchCV
+from sklearn.pipeline import Pipeline
+from sklearn.svm import LinearSVC
+
+
+def score_bounds(cv_results):
+    """
+    Calculate the upper/lower bounds within 1 standard deviation
+    of the best `mean_test_scores`.
+
+    Args:
+        cv_results: see attribute cv_results_ of `GridSearchCV`
+
+    Return:
+        upper/lower bounds within 1 standard deviation of the
+        best `mean_test_score`.
+    """
+    best_test_score = max(cv_results['mean_test_score'])
+    max_score_index = cv_results['mean_test_score'].argmax()
+    # Use the std of `best_test_score` across all cv splits
+    std_test_score = cv_results['std_test_score'][max_score_index]
+    score_upper = np.minimum(best_test_score + std_test_score, 1)
+    score_lower = np.maximum(best_test_score - std_test_score, 0)
+    return score_upper, score_lower
+
+
+def refit_callable(cv_results):
+    """
+    Balance model complexity with cross-validated score.
+
+    Args:
+        cv_results: see attribute cv_results_ of `GridSearchCV`.
+
+    Return:
+        Index of a model that has the fewest PCA components
+        while has test score within 1 standard deviation of the best
+        `mean_test_score`.
+    """
+    test_score_upper, test_score_lower = score_bounds(cv_results)
+    n_components = cv_results['param_reduce_dim__n_components']
+    test_scores = cv_results['mean_test_score']
+    componet_score_lst = list(zip(n_components, test_scores))
```

Comment:
You should do this in a more idiomatic numpy way. Here's how I might do it, but there might be something neater:
```py
best_score_idx = np.argmax(cv_results['mean_test_score'])
threshold = cv_results['mean_test_score'][best_score_idx] - cv_results['std_test_score'][best_score_idx]
candidate_idx = np.flatnonzero(cv_results['mean_test_score'] >= threshold)
best_idx = candidate_idx[cv_results['param_reduce_dim__n_components'][candidate_idx].argmin()]
```

Reply:
:+1: Thanks! I really like the way you select the candidate_idx.